scripts/setup_ubuntu.sh

```diff
@@ -49,25 +51,39 @@ run_as_original_user() {
 # Add custom repositories
 #
 curl -fL https://apt.kitware.com/keys/kitware-archive-latest.asc | apt-key add -
-apt-add-repository -y $NO_UPDATE "deb https://apt.kitware.com/ubuntu/ ${RELEASE} main"
+if [ "$VERSION" == "21" ]; then
+    # no hirsute aka 21.04 apt repo yet, use focal one instead
+    apt-add-repository -y $NO_UPDATE "deb https://apt.kitware.com/ubuntu/ focal main"
+else
+    apt-add-repository -y $NO_UPDATE "deb https://apt.kitware.com/ubuntu/ ${RELEASE_CODENAME} main"
+fi
 
 add-apt-repository -y $NO_UPDATE ppa:git-core/ppa
 
 # Clang isn't present in the xenial repos so we need to add the repo
-VERSION=$(lsb_release --release --short | cut -d . -f 1)
-if [ "$VERSION" == "16" ]; then
+# Clang-10 isn't present in the hirsute repos any more, so we need to add the upstream repo
+if [[ "$VERSION" == "16" ||  "$VERSION" == "21" ]]; then
   curl -fL https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
-  apt-add-repository -y $NO_UPDATE "deb http://apt.llvm.org/${RELEASE}/ llvm-toolchain-${RELEASE}-10 main"
+  apt-add-repository -y $NO_UPDATE "deb http://apt.llvm.org/${RELEASE_CODENAME}/ llvm-toolchain-${RELEASE_CODENAME}-10 main"
 fi
 
 if [[ -n "${SETUP_TOOLCHAIN_VARIANTS}" ]]; then
   apt-add-repository -y $NO_UPDATE ppa:ubuntu-toolchain-r/test
 fi
 
-RELEASE_ID="$(lsb_release --id --short | tr 'A-Z' 'a-z')"
-RELEASE_CODENAME="$(lsb_release --codename --short)"
-curl "https://apache.jfrog.io/artifactory/arrow/$RELEASE_ID/apache-arrow-apt-source-latest-$RELEASE_CODENAME.deb" \
-  --output /tmp/apache-arrow-apt-source-latest.deb
+
+if [ "$VERSION" == "21" ]; then
+    # no hirsute aka 21.04 release of apache arrow yet, use focal one instead
+    # we must also get libre2-5 from focal
+    curl "https://apache.jfrog.io/artifactory/arrow/$RELEASE_ID/apache-arrow-apt-source-latest-focal.deb" \
+         --output /tmp/apache-arrow-apt-source-latest.deb
+    wget "mirrors.kernel.org/ubuntu/pool/main/r/re2/libre2-5_20200101+dfsg-1build1_amd64.deb" \
+         -O /tmp/libre2-5.deb
+    apt install -yq /tmp/libre2-5.deb && rm /tmp/libre2-5.deb
```

Comment:
There is a way to add an older releases repo to apt and then access it using `libre2/focal` as the package descriptor. I would prefer that.

Reply:
something like that can definitely be done with apt .pref files:
1) add the focal apt repo source.list
2) add the focal.pref file, something like:
```
# only allow this package if no other version is installed
Package: libre2-5
Pin: release n=focal
Pin-Priority: 50

# never use other packages from focal
Package: *
Pin: release n=focal
Pin-Priority: -1
```

That said, I'm not convinced that adding the full focal repo and pref file actually results in a cleaner solution vs just grabbing the one package we need and installing it.
We really only need libre2-5 until apache arrow releases a 21.04 package, which will likely depend on libre2-9 instead.
Just grabbing the single package we need makes the future clean up straight forward, just remove libre2-5 (it might even be added to the autoremove list when we install the newer apache arrow, but I'm not 100% confident on that). 

Of course, I'm open to discussion here.

Reply:
Fair. Can you put a comment here if you are making other changes? Because if we start using more than one file from other releases it would be worth setting up a nice way to pull them in.

BTW, I think you are over complicating things. I think you can reasonably safely just add the source line for the old release and then pull in packages using the special version syntax I mentioned. The trick you are showing is only needed if you want to selectively get packages from a NEWER release since those will generally have newer version numbers and hense override the older release by default.

Reply:
I tested this out on my hirsute install. Added the repo with `sudo apt-add-repository "deb http://us.archive.ubuntu.com/ubuntu/ focal main restricted"` which came right from my focal machines sources.list


apt shows we are getting `libre2-5` from focal:
```
apt list libre2-5
Listing... Done
libre2-5/focal 20200101+dfsg-1build1 amd64
libre2-5/focal 20200101+dfsg-1build1 i386
```
but since the package is only available from focal, we don't have to specify the `libre2-5/focal` and apt assumes we just want the only available package instead. 
```
❯ sudo apt install libre2-5
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  libre2-5
0 upgraded, 1 newly installed, 0 to remove and 15 not upgraded.
Need to get 162 kB of archives.
After this operation, 485 kB of additional disk space will be used.
Get:1 http://us.archive.ubuntu.com/ubuntu focal/main amd64 libre2-5 amd64 20200101+dfsg-1build1 [162 kB]
Fetched 162 kB in 0s (476 kB/s)
Selecting previously unselected package libre2-5:amd64.
.....
```
The side effect of this being now any package no longer available in hirsute will get pulled automatically from focal, instead of making the user aware the package is depreciated for their system. This definitely can lead to worse FrankenDebian without the user knowing. I would prefer to require an active step before pulling in old releases packages, which would require the `.perf` file.

Reply:
Valid point. I didn't think about how the package wouldn't exist at all in the newer verion.

The direct download approach is probably best for now.

We eventually need to move away from using all these external repos for dependencies of our packages. Though if some stay around for the build environment it's not a big deal.